src/solve/fulfill.rs

```diff
@@ -115,152 +133,235 @@ impl<'s> Fulfill<'s> {
+        for goal in self.to_refute.drain(..) {
+            let canonicalized = self.infer.canonicalize(&goal);
+            let goal = canonicalized.quantified.value;
+
+            // Negation cannot be used to resolve existential variables, and do
+            // not have a useful (for us) logial meaning when they contain
+            // existential variables; treat negative goals with free variables
+            // as ambiguous
+            if !canonicalized.free_vars.is_empty() {
+                return Ok(Outcome::Incomplete);
             }
+
+
+            // Try to *solve* the goal in a fresh Fulfill
+            let mut fulfill = Fulfill::new(self.solver);
+            fulfill.push_goal(&goal.environment, goal.goal);
+
+            // Negate the result
+            if let Ok(solution) =  fulfill.solve(Substitution::empty()) {
+                match solution {
+                    Solution::Unique(_) => Err("refutation failed")?,
+                    Solution::Ambig(_) => return Ok(Outcome::Incomplete),
+                }
+            }
+        }
+
+        Ok(Outcome::Complete)
+    }
+
+    /// Try to fulfill all pending obligations. The returned solution will
+    /// transform `subst` substitution with the outcome of type inference by
+    /// updating the replacements it provides.
+    pub fn solve(mut self, subst: Substitution) -> Result<Solution> {
+        // We first solve all positive (to_prove) goals, since these are the
+        // only goals that can affect inference state or add new obligations.
+        let positive = self.fulfill_positive()?;
+
+        // Next we look at the goals we must refute; doing so cannot create new
+        // obligations of any form
+        let negative = self.fulfill_negative()?;
+
+        if (positive.is_complete() && negative.is_complete()) && !self.ambiguous {
+            // No obligations remain, so we have definitively solved our goals,
+            // and the current inference state is the unique way to solve them.
+
+            let constraints = self.constraints.into_iter().collect();
+            let constrained = self.infer.canonicalize(&ConstrainedSubst { subst, constraints });
+            return Ok(Solution::Unique(constrained.quantified))
         }
 
-        Ok(solution.successful)
+        // Otherwise, we have (positive or negative) obligations remaining, but
+        // haven't proved that it's *impossible* to satisfy out obligations. we
+        // need to determine how to package up what we learned about type
+        // inference as an ambiguous solution.
+
+        if subst.is_trivial(&mut self.infer) {
+            // In this case, we didn't learn *anything* definitively. So now, we
+            // go one last time through the positive obligations, this time
+            // applying even *tentative* inference suggestions, so that we can
+            // yield these upwards as our own suggestions. In particular, we
+            // yield up the first one we can find.
```

Comment:
**Question:** First in what sense? we're popping the obligations, after all, so this isn't the first we would have encountered had we been applying the substs "in order", right? (Or am I missing something?)

(I made this comment earlier, in my review, but GH marked it as "outdated" for some reason.)